locust/dispatch.py

```diff
@@ -49,548 +48,134 @@ class UsersDispatcher(Iterator):
+        yield
 
-        The assignment also ensures that each worker runs the same amount of users (irrespective of the user class
-        of those users). If the total user count does not yield an integer when divided by the number of workers,
-        then some workers will have one more user than the others.
-        """
-        assigned_users = {
-            worker_node.id: {user_class: 0 for user_class in self._user_classes} for worker_node in self._worker_nodes
-        }
-
-        # We need to copy to prevent modifying `user_classes_count`.
-        user_classes_count = self._user_classes_count.copy()
+        delta = time.perf_counter() - t0_rel
 
-        user_count = sum(user_classes_count.values())
+        self._dispatch_iteration_durations.append(delta)
 
-        # If `remainder > 0`, it means that some workers will have `users_per_worker + 1` users.
-        users_per_worker, remainder = divmod(user_count, len(self._worker_nodes))
-
-        for user_class in self._user_classes:
-            if sum(user_classes_count.values()) == 0:
-                # No more users of any user class to assign to workers, so we can exit this loop.
-                break
+        # print("Dispatch cycle took {:.3f}ms".format(delta * 1000))
 
-            # Assign users of `user_class` to the workers in a round-robin fashion.
-            for worker_node in itertools.cycle(self._worker_nodes):
-                if user_classes_count[user_class] == 0:
-                    break
+        if self._current_user_count == self._target_user_count:
+            # No sleep when this is the last dispatch iteration
+            return
 
-                number_of_users_left_to_assign = user_count - self._number_of_assigned_users_across_workers(
-                    assigned_users
-                )
+        sleep_duration = max(0.0, self._wait_between_dispatch - delta)
+        gevent.sleep(sleep_duration)
 
-                if (
-                    self._number_of_assigned_users_for_worker(assigned_users, worker_node) == users_per_worker
-                    and number_of_users_left_to_assign > remainder
-                ):
+    def _ramp_up(self) -> Dict[str, Dict[str, int]]:
+        initial_user_count = self._current_user_count
+        for user in self._user_gen_generator:
+            worker_name = next(self._worker_node_id_generator)
+            self._users_on_workers[worker_name][user] += 1
+            self._current_user_count += 1
+            if self._current_user_count >= min(
+                initial_user_count + self._user_count_per_dispatch_iteration, self._target_user_count
+            ):
+                return deepcopy(self._users_on_workers)
+
+    def _ramp_down(self) -> Dict[str, Dict[str, int]]:
+        initial_user_count = self._current_user_count
+        for user in self._user_gen_generator:
+            while True:
+                worker_name = next(self._worker_node_id_generator)
```

Comment:
does this ensure that if you ramp from for example 10->20->10 users you get back the distribution you originally had? (that was what I was getting at with my stopped_users list)

Reply:
and does it also preserve the distribution if a worker disconnects and users have to be spawned across the other workers?

Reply:
> does this ensure that if you ramp from for example 20->10->20

Yes, I added the `active_users` list in commits made afterward and there's a test specifically to ensure this use case works.

> and does the algorithm also preserve the distribution if a worker disconnects and users have to be spawned across the remaining workers?

Absolutely. Again, the actual logic was added in commits made after your comment.